src/server/frontend_wayland/wl_touch.cpp

```diff
@@ -79,29 +71,21 @@ void mf::WlTouch::handle_event(MirTouchEvent const* touch_ev, WlSurface* surface
         case mir_touch_action_change:
         {
             auto const transformed = surface->transform_point(point);
-            if (transformed)
+            if (focused_surface_for_ids[touch_id] == transformed.surface)
```

Comment:
This seems wrong (although the existing code *also* seems wrong!). This is going to ensure that a (sub)surface never receives a touch event outside its bounds?

I believe the correct behaviour is for the surface that received the `touch_down` to receive all further events for that touch until `touch_up`. I don't know how that interacts with the input region, however...

Reply:
I believe you are correct. Let me see what I can do about this.

Reply:
Lemme write some tests, so the code I end up with actually works.